File: Excel/440 Challenge.py
import pandas as pd
import numpy as np

test = pd.read_excel('440 List of Numbers Expressed as Sum of Two Squares.xlsx', usecols='A', nrows=30)

def is_sum_of_squares(x):
    squares = np.arange(1, int(np.sqrt(x)) + 1) ** 2
    for i in range(len(squares)):
        for j in range(i+1, len(squares)):
            if squares[i] + squares[j] == x:
                return True
    return False

result = pd.DataFrame(np.arange(1, 101), columns=['Number']).astype("int64")
result['Is_Sum_Of_Squares'] = result['Number'].apply(is_sum_of_squares)
result = result[result['Is_Sum_Of_Squares'] == True]
result = result.drop(columns='Is_Sum_Of_Squares').reset_index(drop=True)

print(result['Number'].equals(test['Answer Expected'])) # True

File: Excel/400-499/440/440 Challenge.R
library(tidyverse)
library(readxl)

test = read_excel("Excel/440 List of Numbers Expressed as Sum of Two Squares.xlsx", range = "A1:A30")

is_sum_of_squares = function(x) {
  squares = (1:floor(sqrt(x)))^2
  any(map_lgl(squares,  ~ any(x == .x + squares[squares != .x])))
}

result = data.frame(numbers = 1:100 %>% as.numeric()) %>%
  filter(map_lgl(numbers, is_sum_of_squares))

identical(result$numbers, test$`Answer Expected`)
# [1] TRUE
